replica/database.cc

```diff
@@ -3184,4 +3184,112 @@ database::on_effective_service_levels_cache_reloaded() {
+        auto t = local_db.get_tables_metadata().get_table_if_exists(id);
+        if (!t) {
+            throw std::runtime_error(fmt::format("sample_data_files: table {} does not exist", id));
+        }
+        local_state->schema = t->schema();
+        local_state->snapshot = co_await t->take_sstable_set_snapshot();
+        size_t my_total_chunks = 0;
+        for (const auto& sst : local_state->snapshot) {
+            my_total_chunks += sst.sst->data_size() / chunk_size;
+        }
+        local_state->local_chunks = my_total_chunks;
+    }));
+
+    uint64_t total_chunks = 0;
+    for (const auto& x : state) {
+        total_chunks += x->local_chunks;
+        x->cross_shard_offset = total_chunks;
+    }
+
+    if (total_chunks == 0) {
+        co_return utils::chunked_vector<bytes>{};
+    }
+
+    const auto chosen_chunks = compute_random_sorted_ints(total_chunks - 1, n_chunks);
+
+    auto sample_one_shard = [
+        &state = std::as_const(state),
+        &chosen_chunks = std::as_const(chosen_chunks),
+        chunk_size
+    ] (database& local_db) -> future<utils::chunked_vector<bytes>> {
+        auto ticket = get_units(local_db._sample_data_files_local_concurrency_limiter, 1);
+
+        auto& local_state = state[this_shard_id()];
+
+        size_t offset = this_shard_id() == 0 ? 0 : state[this_shard_id() - 1]->cross_shard_offset;
+        SCYLLA_ASSERT(local_state->cross_shard_offset >= offset);
+        auto chosen_it = std::ranges::lower_bound(chosen_chunks, offset);
+        const auto chosen_end = std::ranges::lower_bound(chosen_chunks, local_state->cross_shard_offset);
+        auto sst_it = local_state->snapshot.begin();
+        uint64_t local_chosen_chunks = chosen_end - chosen_it;
+
+        auto get_next_chunk = [&] () -> std::pair<sstables::shared_sstable, uint64_t> {
+            SCYLLA_ASSERT(sst_it != local_state->snapshot.end());
+            while (*chosen_it - offset >= sst_it->sst->data_size() / chunk_size) {
+                offset += sst_it->sst->data_size() / chunk_size;
+                ++sst_it;
+                SCYLLA_ASSERT(sst_it != local_state->snapshot.end());
+            }
+            SCYLLA_ASSERT(chosen_it != chosen_end);
+            auto offset_in_chunks = *chosen_it++;
+            return {sst_it->sst, (offset_in_chunks - offset) * chunk_size};
+        };
+
+        utils::chunked_vector<bytes> result;
+        result.reserve(local_chosen_chunks);
+        int concurrency_limit = 10;
+        co_await max_concurrent_for_each(std::views::iota(uint64_t(0), local_chosen_chunks), concurrency_limit, coroutine::lambda([&] (int) -> future<> {
+            auto permit = co_await local_db._system_read_concurrency_sem.obtain_permit(
+                local_state->schema, "sample_data_files", chunk_size, no_timeout, nullptr);
+            auto [sst, offset] = get_next_chunk();
+            auto buf = co_await sst->data_read(offset, chunk_size, permit);
```

Comment:
this might enable unnecessary read ahead since data_read() unconditionally uses read ahead of 4 (although it should be reduced to 1 by file input stream even though slow start is not enabled due to lack of history with data_read()) .
Read ahead of 1 means emitting an additional read that will be discarded, so unnecessary read ampl.

Another user of sstable::data_read() is reversed partition read, and it also seems to be affected (it could actually benefit from reversed read ahead but it's not supported AFAIK). Since data_read() is not an actual streaming reader, just consumes a range in one go, I think we should just patch it to disable read ahead altogether.

Reply:
> this might enable unnecessary read ahead

No, because we are passing a finite length (`chunk_size`), and the file stream won't read beyond that length, regardless of what the read ahead is.

I.e. if I ask Seastar to read a file stream from the range [0x1000, 0x2000) of a file with a read ahead of 0x100000, Seastar's is supposed to read [0x1000, 0x2000), not [0x1000, 0x100000].